### ISSUE TITLE
Cyclic structures

### ISSUE THREAD

**Issue by juangamnik** [gk-brown/HTTP-RPC#44]
Why don't you support cyclic structures?:

> Nested structures are supported, but reference cycles are not permitted.
[HTTP-RPC documentation](https://github.com/gk-brown/HTTP-RPC)

You could use [JSOG](https://github.com/jsog/jsog) for dereferencing. Not allowing cyclic structures (like a student with a bidirectional many-to-many association to professors) is a very hard restriction for a RPC-framework.

**Comment by gk-brown** [comment]
Great question! The short answer is simplicity. In general, I've tried to avoid overloading the framework with features that could reasonably be handled at the application level.

However, I wasn't aware of JSOG. It's an interesting concept. It would require some additional processing to detect/reconstruct cycles as well as some additional bandwidth for the ID/ref data, but I can see the utility. 

I'm not sure I'd want to automatically apply it to every response, but it might make sense as an opt-in feature. For example, if I know a method might returns a cyclic structure, I could annotate it and trigger JSOG handling of the result.

**Comment by juangamnik** [comment]
It may even reduce bandwidth and increase the referential integrity of an object tree (and an object graph too for sure) since objects that occur more than once in a request or a response are written only once and referenced by id at every other occurrence. For the resulting object graph is JSOG completely transparent. There is a Jackson extension library for JSOG, too (I don't know which library you use on the server side):

```xml
<dependency>
  <groupId>com.voodoodyne.jackson.jsog</groupId>
  <artifactId>jackson-jsog</artifactId>
  <version>1.1</version>
</dependency>
```

**Comment by gk-brown** [comment]
I wrote the JSON serialization code myself, so adding it to the server wouldn't be difficult. I also wrote the deserialization logic for the Java client, so that wouldn't be a problem either. But the iOS client uses `NSJSONSerialization` and the JavaScript client uses `JSON.parse()`, and I don't think either of those currently support JSOG.

That said, it might be OK to support JSOG on the server only in the short term and add client support later, if it makes sense. This implies that any JSOG-enabled server could only be accessed by a JSOG-aware client library, which, in this case, would need to be a 3rd-party library such as jackson-jsog, since the HTTP-RPC client libraries won't support it.

I'm still not sure what the best approach would be for enabling JSOG response processing, though. It could be done on a per-method basis via an annotation, at the service level via a configuration parameter, or even at the JVM level via a system property. Alternatively, it could be managed via a custom request header indicating that the client is JSOG-capable. Any thoughts?

**Comment by juangamnik** [comment]
I think you can do the JSOG-stuff in the iOS client in a pre-/post processing step (transparently). JSOG is more or less a *one-hundred-liner*. It is quite straightforward.

But if you want to be compatible to arbitrary clients there, will be a need for some kind of *compatibility mode*. JAX-RS for example offers `@Consume` and `@Produce` annotations on a per method (perhaps even per class, I'm unsure) basis where you can configure whether your server provides JSON or XML serialization (or both) and switches between the configured JSON (e.g. Jackson) and XML (e.g. JAXB) serialization implementation. The client may choose between them via the corresponding `Accept`-Header `text/xml` and `application/json`. Unfortunately, there is no MIME-Type for JSOG in the *IANA* registry of IETF.

A possible solution could be to send [two MIME-types](http://www.w3.org/Protocols/HTTP/HTRQ_Headers.html) `application/json` and `application/jsog+json`. or two use a custom header as you said.

**Comment by gk-brown** [comment]
Yes, JSOG could be implemented via post-processing. And yes, there needs to be some sort of compatibility mode - that's what I meant by "enabling" JSOG. Not all clients will want or need it, so there needs to be a way to toggle it on/off. 

I'm currently thinking that a header might be the best approach, since it would allow the caller to opt in. However, if a caller failed to request JSOG on a call that returns a cyclic data structure, it would result in an infinite loop on the server. So perhaps the service would need some way to flag a method as requiring JSOG as well. If a caller invokes a JSOG-required method but fails to specify the JSOG header, the server could simply return an error response.

**Comment by juangamnik** [comment]
> if a caller failed to request JSOG on a call that returns a cyclic data structure, it would result in an infinite loop on the server

I think this cannot be decided completely upfront (at compilation time), so that flagging the method is IMHO inappropriate. In your current implementation an object to be returned could contain cycles in a reference either. So, independently from the support of cyclic structures it would be a good thing to recognize cycles during serialization and (if JSOG is disabled) raise a runtime exception. JSOG could easily be adapted to have a mode where it detects cycles (like a *mark* mechanism in a breadth-first search) instead of dereferencing.

**Comment by gk-brown** [comment]
That's true - the current implementation doesn't detect cycles and will go into an infinite loop if one is found. So detecting cycles at runtime rather than providing hints at compile time is probably a better approach.

**Comment by gk-brown** [comment]
Closing. The server component has been retired, so this issue no longer applies.

**Comment by juangamnik** [comment]
Retired?

**Comment by gk-brown** [comment]
Right. The project doesn't seem to be getting as much traction as I had hoped, so I was planning to phase out the server component and focus on the client. Are you currently using it? I really don't have any insight into usage. I opened up a Gitter channel a few weeks ago hoping to spur some discussion, but no one has posted anything yet. I tried something similar with Disqus a few months back - same result. 

I've been thinking about restoring server support in a future update, in a slightly different but mostly compatible form. If enough people are actually using it, I'd consider doing that sooner rather than later. 

For what it's worth, I moved all of the template stuff to a new project (everything is still in the development branch since it's not quite ready for release):

https://github.com/gk-brown/JTemplate

If I do update the server, it will most likely include this project as a dependency.
